File: 2023-04-14/F/F.go
// https://codeforces.com/problemset/problem/1343/B

// CodeForces 1343B - Balanced Array

// You are given a positive integer n, it is guaranteed that n is even (i.e. divisible by 2).
// You want to construct the array a of length n such that:
//  • The first n/2 elements of a are even (divisible by 2);
//  • the second n/2 elements of a are odd (not divisible by 2);
//  • all elements of a are distinct and positive;
//  • the sum of the first half equals to the sum of the second half (∑_(i=1)^(n/2)(ai)=∑_(i=n/2+1)^(n)(ai)).
// If there are multiple answers, you can print any. It is not guaranteed that the answer exists.
// You have to answer t independent test cases.

// Input
// The first line of the input contains one integer t (1≤t≤10^4) — the number of test cases. Then t test cases follow.
// The only line of the test case contains one integer n (2≤n≤2⋅10^5) — the length of the array. It is guaranteed that that n is even (i.e. divisible by 2).
// It is guaranteed that the sum of n over all test cases does not exceed 2⋅10^5 (∑n≤2⋅10^5).

// Output
// For each test case, print the answer — "NO" (without quotes), if there is no suitable answer for the given test case or "YES" in the first line and any suitable array a1,a2,…,an (1≤ai≤10^9) satisfying conditions from the problem statement on the second line.

// Example
// input
// 5
// 2
// 4
// 6
// 8
// 10
// output
// NO
// YES
// 2 4 1 5
// NO
// YES
// 2 4 6 8 1 3 5 11
// NO

// Tutorial
// Firstly, if n is not divisible by 4 then the answer is "NO" because the parities of halves won't match. Otherwise, the answer is always "YES". Let's construct it as follows: firstly, let's create the array [2,4,6,…,n,1,3,5,…,n−1]. This array is almost good except one thing: the sum in the right half is exactly n/2 less than the sum in the left half. So we can fix it easily: just add n/2 to the last element.

package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	var reader = bufio.NewReader(os.Stdin)
	var writer = bufio.NewWriter(os.Stdout)
	var c int
	var t, n, i uint

	for c, _ = fmt.Fscan(reader, &t); c == 1; c, _ = fmt.Fscan(reader, &t) {
		for ; t > 0; t-- {
			fmt.Fscan(reader, &n)

			n /= 2

			if n&1 == 0 {
				fmt.Fprintln(writer, "YES")

				for i = 1; i <= n; i++ {
					fmt.Fprint(writer, 2*i, " ")
				}

				for i = 1; i < n; i++ {
					fmt.Fprint(writer, 2*i-1, " ")
				}

				fmt.Fprintln(writer, 3*n-1)
			} else {
				fmt.Fprintln(writer, "NO")
			}
		}
	}

	writer.Flush()
}

File: 2023-05-26/F/F.cpp
// https://onlinejudge.org/index.php?option=com_onlinejudge&Itemid=8&category=441&page=show_problem&problem=4027

// UVA 12582 - Wedding of Sultan

// As usual Sultan Mahmud is very busy. He works days and nights at office. If you ask him, “Sultan,
// which day of the week is this?” He will look at you for a while and say, “I think I have 3 more days
// till deadline!” But one day the scenario changed after receiving a call. He usually ignores phone calls
// from everyone (even from his fiance) but this time he couldn’t ignore because of the importance of the
// person! This person was his to-be mother-in-law. So he received the call and heard, “Son, only 30
// days left of your wedding ceremony, so I am sending a tailor for the measurement for your suit.” Sultan
// now remembered, he is about to get married but looking at thyself, he got surprised! When did he get
// so fat! “Umm.. Mom can it be arranged 10 days later?” He wants to buy some time so that he can
// exercise and lose extra weight. So he immediately went out with his bicycle to the large garden beside
// his house.
// There are several trails in the garden. A trail starts from one water sprinkler to another and the
// sprinkler are marked by distinct letters from ‘A’ to ‘Z’. The trails are designed in such a way that from
// the sprinkler at entrance you can go to any other sprinkler using exactly one path if you do not traverse
// a trail more than once.
// While traversing the trails with his cycle, Sultan notes the names of the sprinklers in his notepad.
// He will write down the name of a sprinkler if he enters the sprinkler for the first time or leaves this
// sprinkler for the last time. And not surprisingly, geek Sultan follows a peculiar method to ensure that
// he visits all the trails of the garden. When he comes to a sprinkler he looks for a trail which he has not
// traversed yet. If he finds such trail, he follows that one. Otherwise, he uses the trail that he used to
// come here for the first time except if it’s the entrance he stops exercising. He always starts from the
// entrance and guess what, his peculiar strategy always guarantees to finish him at the entrance and all
// the trails are also visited.
// For example, in the above garden the main entrance is at A. So Sultan will start from A. When
// Sultan is at A, he can choose either of the trails. Say he chooses the trail leading to E. Then he can
// choose the trail to G or trail to F. Say he chooses F. Now he does not have any unvisited trail from F,
// so he will go back to E. Now he must choose trail to G and then similarly will come back to E and back
// to A. Then he will go towards B. Now he again has two choices. He can go to C or D, say he goes to
// C, then he will be back to B, then will go to D, and hence back to B and also back to A thus finishing
// his exercise. So after his exercise you will find: AEFFGGEBDDCCBA in his notepad. Can you find
// the number of trails attached to the sprinklers just looking at the sequence written in the notepad?

// Input

// First line of the test file contains a positive integer T (T ≤ 100) denoting the number of test cases.
// Hence follows T lines, each containing a valid sequence of sprinkler names. A sprinkler name will
// always be capital Latin letter (‘A’, ‘B’, . . . , ‘Z’). You may assume that there will be at least two sprinklers
// in garden, otherwise there would have been no meaning of exercise right?

// Output

// For each case output the case number in the first line, followed by the number of trails for each sprinkler.
// First print the sprinkler name followed by the count of trails. These lines should be in lexicographical
// order of sprinkler name. Note that, you should not print about a sprinkler which is not present in the
// garden. Look at the sample input output for more specific format of input output.

// Sample Input

// 2
// AEFFGGEBDDCCBA
// ZAABBZ

// Sample Output

// Case 1
// A = 2
// B = 3
// C = 1
// D = 1
// E = 3
// F = 1
// G = 1
// Case 2
// A = 1
// B = 1
// Z = 2

// given graph DFS traversal, count the degree of each vertex

#include <iostream>
#include <map>
#include <stack>
#include <string>

int main() {
    unsigned t, c, i;
    std::string str;
    std::stack<char> s;
    std::map<char, unsigned> degree;
    
    while (std::cin >> t) {
        for (c = 1; c <= t; c++) {
            std::cin >> str;
            s = std::stack<char>();
            degree.clear();

            s.push(str[0]);

            for (i = 1; i < str.size(); i++) {
                if (s.top() == str[i]) {
                    s.pop();
                } else {
                    degree[s.top()]++;
                    degree[str[i]]++;
                    s.push(str[i]);
                }
            }

            std::cout << "Case " << c << "\n";

            for (std::map<char, unsigned>::iterator it = degree.begin();
                 it != degree.end();
                 it++) {
                std::cout << it->first << " = " << it->second << "\n";
            }
        }
    }

    return 0;
}
